dataclients/kube/kube.go

```diff
@@ -0,0 +1,303 @@
+	}
+
+	if p, ok := port.number(); ok {
+		log.Debugf("service port as number: %d", p)
+		return fmt.Sprintf("http://%s:%d", s.Spec.ClusterIP, p), nil
+	}
+
+	pn, _ := port.name()
+	for _, pi := range s.Spec.Ports {
+		if pi.Name == pn {
+			log.Debugf("service port found by name: %s -> %d", pn, pi.Port)
+			return fmt.Sprintf("http://%s:%d", s.Spec.ClusterIP, pi.Port), nil
+		}
+	}
+
+	log.Debugf("service port not found by name: %s", pn)
+	return "", errServicePortNotFound
+}
+
+// TODO: find a nicer way to autogenerate route IDs
+func routeID(namespace, name, host, path string) string {
+	namespace = nonWord.ReplaceAllString(namespace, "_")
+	name = nonWord.ReplaceAllString(name, "_")
+	host = nonWord.ReplaceAllString(host, "_")
+	path = nonWord.ReplaceAllString(path, "_")
+	return fmt.Sprintf("kube_%s__%s__%s__%s", namespace, name, host, path)
+}
+
+// converts the default backend if any
+func (c *Client) convertDefaultBackend(i *ingressItem) (*eskip.Route, bool, error) {
+	// the usage of the default backend depends on what we want
+	// we can generate a hostname out of it based on shared rules
+	// and instructions in annotations, if there are no rules defined
+
+	// this is a flaw in the ingress API design, because it is not on the hosts' level, but the spec
+	// tells to match if no rule matches. This means that there is no matching rule on this ingress
+	// and if there are multiple ingress items, then there is a race between them.
+	// TODO: don't crash when no Spec
+	if i.Spec.DefaultBackend == nil {
+		return nil, false, nil
+	}
+
+	address, err := c.getServiceAddress(
+		i.Metadata.Namespace,
+		i.Spec.DefaultBackend.ServiceName,
+		i.Spec.DefaultBackend.ServicePort,
+	)
+
+	if err != nil {
+		return nil, false, err
+	}
+
+	r := &eskip.Route{
+		Id:      routeID(i.Metadata.Namespace, i.Metadata.Name, "", ""),
+		Backend: address,
+	}
+
+	return r, true, nil
+}
+
+func (c *Client) convertPathRule(ns, name, host string, prule *pathRule) (*eskip.Route, error) {
+	if prule.Backend == nil {
+		return nil, fmt.Errorf("invalid path rule, missing backend in: %s/%s/%s", ns, name, host)
+	}
+
+	address, err := c.getServiceAddress(ns, prule.Backend.ServiceName, prule.Backend.ServicePort)
+	if err != nil {
+		return nil, err
+	}
+
+	var pathExpressions []string
+	if prule.Path != "" {
+		pathExpressions = []string{prule.Path}
+	}
+
+	r := &eskip.Route{
+		Id:          routeID(ns, name, host, prule.Path),
+		PathRegexps: pathExpressions,
+		Backend:     address,
+	}
+
+	return r, nil
+}
+
+// logs if invalid, but proceeds with the valid ones
+// should report failures in Ingress status
+//
+// TODO:
+// - check how to set failures in ingress status
+func (c *Client) ingressToRoutes(items []*ingressItem) []*eskip.Route {
+	routes := make([]*eskip.Route, 0, len(items))
+	for _, i := range items {
+		if i.Metadata == nil || i.Metadata.Namespace == "" || i.Metadata.Name == "" {
+			log.Errorf("invalid ingress item: missing metadata")
```

Comment:
not sure what your conventions say: but I would log this as warning and not error (as we will happily continue with the next item)